# Importing libraries

In [10]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [11]:
# Turning project folder into string path 
path = r'C:\Users\keirr\11-2020 Instacart Basket Analysis'

In [12]:
# Importing products dataset with path
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'OG Data', 'products.csv'))

In [13]:
# Importing orders_wrangled dataset as df_ords with path
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [14]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_the_week  order_hour_of_day  days_since_last_order  
0                             2                  8                    NaN  
1                             3                  7                   15.0  
2                             3                 12                   21.0  
3                             4                  7                   29.0  
4                             4                 15                   28.0  
...                         ...                ...                    ...  
3421078                       5                 18                   29.0  
3421079                       4                 10                   30.0  
3421080                       1                 12                   18.0  
3421081                       1                 12                    7.0  
3421082                       6                 14                   30.0  

[3421083 rows x 8 columns]

# Contents
## 1. Dropping 'Unnamed: 0' column
## 2. Data Consistency Checks: Products dataset
## 3. Data Consistency Checks: Orders dataset
## 4. Exporting Orders dataset

# 1. Dropping 'Unnamed: 0' column

In [15]:
# Issue with import of df_ords.csv, dropping 'unnamed' column
df_ords.drop('Unnamed: 0', axis=1, inplace=True)

In [16]:
# Checking df_ords
df_ords

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0         2539329        1    prior             1                       2   
1         2398795        1    prior             2                       3   
2          473747        1    prior             3                       3   
3         2254736        1    prior             4                       4   
4          431534        1    prior             5                       4   
...           ...      ...      ...           ...                     ...   
3421078   2266710   206209    prior            10                       5   
3421079   1854736   206209    prior            11                       4   
3421080    626363   206209    prior            12                       1   
3421081   2977660   206209    prior            13                       1   
3421082    272231   206209    train            14                       6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 7 columns]

In [17]:
# Dropping eval_set column from df_ords
df_ords = df_ords.drop(columns = ['eval_set'])

In [18]:
# One more check of df_ords before task at end
df_ords

order_id  user_id  order_number  orders_day_of_the_week  \
0         2539329        1             1                       2   
1         2398795        1             2                       3   
2          473747        1             3                       3   
3         2254736        1             4                       4   
4          431534        1             5                       4   
...           ...      ...           ...                     ...   
3421078   2266710   206209            10                       5   
3421079   1854736   206209            11                       4   
3421080    626363   206209            12                       1   
3421081   2977660   206209            13                       1   
3421082    272231   206209            14                       6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 6 columns]

# 2. Data Consistency Checks: Products dataset

### Mixed Type Data

In [19]:
# Creating a test df
df_test = pd.DataFrame()

# Creating mix-type columns
df_test['mix'] = ['a', 'b', 1, True]

In [20]:
# Taking a look at created df
df_test

mix
0     a
1     b
2     1
3  True

In [21]:
# Searching df_test for mixed-type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [22]:
# Addressing mix-type data 
df_test['mix'] = df_test['mix'].astype('str')

### Missing Values in Data

In [23]:
# Searching for missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [24]:
# Creating subset for nan values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [25]:
# Checking subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [26]:
# Checking shape of df_prods
df_prods.shape

(49693, 5)

In [27]:
# Addressing missing values by excluding them from new df
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [28]:
# Checking shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

### Duplicates in Data

In [29]:
# Searching for duplicates within subset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [30]:
# Checking subset df_dups
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [31]:
# Checking shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [32]:
# Addressing duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
# Checking shape of df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

In [34]:
# Exporting df_prods_clean_no_dups with 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 3. Data Consistency Checks: Orders dataset

In [35]:
# Investigating data with df.describe()
df_ords.describe()

order_id       user_id  order_number  orders_day_of_the_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06            3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01            2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01            2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00            0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00            1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01            3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01            5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02            6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

### Insights: min/max values don't seem to make sense with knowing the data that is within df_ords; 'order_hour_of_day' shows max of 2.3 (24hrs in a day), 'days_since_last_order' shows max of 3 (30 days I believe is the max)

In [36]:
# Checking for mix-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Insights: no mixed type data found within df_ords, yay!

In [38]:
# Checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_last_order     206209
dtype: int64

### Insights: 206,209 rows in 'days_since_last_order', based on the data in the column perhaps these are instances of one time orders, with no follow up order. Although I could just create new df omitting the data with nan values, marketing/sales team wants to see behavior patterns of customers so this would not be beneficial. 

### Insights: Imputation isn't the best option, as these are most likely one-time customer and it will skew data. Data will be left within column

In [41]:
# Searching for duplicates in df_ords
df_dups = df_ords[df_ords.duplicated()]

In [42]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

### Insights: I am not seeing any duplicates within df_ords

## 4. Exporting Orders dataset

In [43]:
# Exporting df_ords with path 
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))